File: 02 백준/03 그래프 탐색/Python/Prob4179.py

"""
title  : 4179. 불! (Gold 4)
time   : 1144ms     520ms
memory : 39452KB    164816KB
"""

import sys
from collections import deque

def solution():

    input = sys.stdin.readline
    
    R, C = map(int, input().split())    # 행 R, 열 C
    graph = [list(input()) for _ in range(R)]

    return search(graph, R, C)
    
def search(graph, R, C):

    dir = [[1, 0], [0, 1], [-1, 0], [0, -1]]    # 사방탐색

    fire = deque()
    jihun = deque()

    # 환경 세팅
    for i in range(R):
        for j in range(C):
            if graph[i][j] == 'F':
                fire.append((i, j))
            elif graph[i][j] == 'J':
                jihun.append((i, j, 0))

    while jihun:

        # 불 이동
        for _ in range(len(fire)):

            nowr, nowc = fire.popleft()

            for d in range(4):
                nr = nowr + dir[d][0]
                nc = nowc + dir[d][1]

                # 이동 불가능
                if nr < 0 or nc < 0 or nr >= R or nc >= C or graph[nr][nc] == '#' or graph[nr][nc] == 'F':
                    continue
                
                graph[nr][nc] = 'F'
                fire.append((nr, nc))

        # 지훈이 이동
        for _ in range(len(jihun)):

            nowr, nowc, dis = jihun.popleft()

            if nowr == 0 or nowr == R-1 or nowc == 0 or nowc == C-1:
                return dis+1

            for d in range(4):
                nr = nowr + dir[d][0]
                nc = nowc + dir[d][1]

                # 이동 불가능
                if graph[nr][nc] == '#' or graph[nr][nc] == 'F' or graph[nr][nc] == 'J':
                    continue

                graph[nr][nc] = 'J'
                jihun.append((nr, nc, dis+1))

    return 'IMPOSSIBLE'

print(solution())

File: 02 백준/03 그래프 탐색/Java/src/Prob4179.java

/**
* title  : 4179. 불! (Gold 3)
* time   : 604ms
* memory : 70408KB
*/

import java.io.*;
import java.util.*;

public class Prob4179 {
    
    static int R, C;

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        char[][] board = new char[R][C];

        Queue<int[]> ji = new LinkedList<>();
        Queue<int[]> fire = new LinkedList<>();

        for (int i = 0; i < R; i++) {
            String temp = br.readLine();
            for (int j = 0; j < C; j++) {
                board[i][j] = temp.charAt(j);

                if (board[i][j] == 'J') {
                    ji.add(new int[] {i, j, 1});
                } else if (board[i][j] == 'F') {
                    fire.add(new int[] {i, j});
                }
            }
        }

        int answer = bfs(board, ji, fire);

        if (answer == -1) {
            System.out.println("IMPOSSIBLE");
        } else {
            System.out.println(answer);
        }
    }

    public static int bfs(char[][] board, Queue<int[]> ji, Queue<int[]> fire) {

        int[] dr = {0, 0, -1, 1};
        int[] dc = {1, -1, 0, 0};

        int length = 0;
        while (!ji.isEmpty()) {

            length = fire.size();
            for (int i = 0; i < length; i++) {

                int[] now = fire.poll();

                for (int d = 0; d < 4; d++) {
                    int nr = now[0] + dr[d];
                    int nc = now[1] + dc[d];

                    if (checkBoundary(nr, nc) && (board[nr][nc] == 'J' || board[nr][nc] == '.')) {
                        board[nr][nc] = 'F';
                        fire.add(new int[] {nr, nc});
                    }
                }
            }

            length = ji.size();
            for (int i = 0; i < length; i++) {

                int[] now = ji.poll();

                if (now[0] == 0 || now[1] == 0 || now[0] == (R-1) || now[1] == (C-1)) {
                    return now[2];
                }

                for (int d = 0; d < 4; d++) {
                    int nr = now[0] + dr[d];
                    int nc = now[1] + dc[d];
                    
                    if (checkBoundary(nr, nc) && board[nr][nc] == '.') {
                        board[nr][nc] = 'J';
                        ji.add(new int[] {nr, nc, now[2]+1});
                    }
                }
            }
        }

        return -1;
    }

    public static boolean checkBoundary(int r, int c) {
        if (r < 0 || r >= R || c < 0 || c >= C) return false;
        return true;
    }
}
